[BugFix] - Prefetching on GPU is actually executed on CPU

DEKHTIARJonathan opened the pull request:
When applying the following:

```python
dataset = dataset.apply(
    tf.data.experimental.prefetch_to_device(
        '/gpu:0',
        buffer_size=1,
    )
)
```

Tensorflow behind the hood does two additional things:

1. Create an _OptionsDataset (everything is fine since the underlying `_variant_tensor` is not touched)
  - https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/prefetching_ops.py#L76-L78
  - https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/ops/dataset_ops.py#L4323

2. Prefetch a number of batches on GPU depending on buffer_size => **Issue is Here**
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/prefetching_ops.py#L54

The problem is that `gen_dataset_ops.prefetch_dataset` defaults on CPU and basically re-copy the dataset back on the CPU. Effectively killing the purpose and adding some overhead.

The fix is simple, just adding `with ops.device(input_dataset._variant_tensor.device):` depending on where the `_variant_tensor` is already located

CC: @sanjoy 

**To Be Noted:** This issue will re-appear if the user does not prefetch to device as the "last step" of the TF Data Pipeline. Maybe a warning should be executed in case the user does not respect this heuristic

**Eventual Question:** Could someone explain the rationale behind these lines, why prefetching to device should have optimizations and autotune deactivated? 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/prefetching_ops.py#L73-L75

```python
options = dataset_ops.Options()
options.experimental_optimization.apply_default_optimizations = False
options.experimental_optimization.autotune = False
return _CopyToDeviceDataset(
  dataset, target_device=target_device,
  source_device=source_device).with_options(options)
```

Commit:
Allow Prefetching on GPU

sanjoy commented:
CC @jsimsa @mrry @rohan100jain

jsimsa commented:
@DEKHTIARJonathan thank you for the detailed information. Would you be interested in submitting a fix for this issue? If so, please also include a unit test that fails prior to your bug fix and passes afterwards.

As for your eventual question, the reason for those settings is to avoid chaining the `ModelDataset` and `OptimizeDataset` transformations to the end of the input pipeline because the transformations only have GPU kernels.

DEKHTIARJonathan commented:
@jsimsa It's a PR so you just have to merge and we are good ;)

> As for your eventual question, the reason for those settings is to avoid chaining the ModelDataset and OptimizeDataset transformations to the end of the input pipeline because the transformations only have CPU kernels.

Stop me if it doesn't make any sense

Why instead of doing:
1. CopyToDevice
2. Deactivate Optimization & Autotune
3. Prefetch

We don't do:
1. Activate Default Opt & Autotune
2. CopyToDevice
3. Prefetch

jsimsa commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -4266,11 +4266,14 @@ def __init__(self, input_dataset, buffer_size, slack_period=None):
-        buffer_size=self._buffer_size,
-        slack_period=slack_period,
-        **self._flat_structure)
+   
```
Please move this device scope to `prefetching_ops.py`, `prefetch_to_device` implementation.

DEKHTIARJonathan replied:
Not sure it's doable actually. And i think it makes more sense the closer to the prefetching OP as possible.

For instance if someone is reimplementing (for any reason) `def prefetch_to_device(device, buffer_size=None):`, then at least like it is currently implemented we are still good

jsimsa reviewed (changes requested):
Thank you. Please add a unit test to `tensorflow/python/data/experimental/kernel_tests/prefetch_to_device_test.py` that would fail prior to your change but is passing now. You can use the [following ops](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/kernel_tests/scan_test.py#L275-L280) to inspect the device placement of an op during the test.

jsimsa commented:
Sorry, I thought this is an issue (as opposed to a PR) :-).

The options need to be set because otherwise tf.data will chain the two datasets to the end of the input pipeline when creating an iterator for it. However, you have a good point in that the host dataset should have optimizations and autotuning applied to it. This can be done [as follows](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/kernel_tests/scan_test.py#L278-L279) and you are welcome to add this functionality to this PR. Thanks!

jsimsa commented:
I would be in favor of cherrypicking this into TF 2.2.

Commit:
Unittest Added to reproduce the bug

DEKHTIARJonathan commented:
@jsimsa I have added two lines in the unittests, that should be fine and sufficient to show pass/fail.

Btw. I just found that this code lead again to the same issue:
```python
iterator = dataset_ops.make_initializable_iterator(device_dataset)
next_element = iterator.get_next()
```

Commit:
Improved Unittest

Commit:
iterator.get_next() colocation bug also fixed

DEKHTIARJonathan commented:
@jsimsa I pushed the unittest showing both issues and fixed the problem in the same way for prefetch and iterator.get_next()

Please let me know if I can do anything else

jsimsa reviewed (changes requested):
The right way to fix placement issues is to update the TensorFlow placer implementation but such changes would require much more testing as they have the potential to affect a wider range of users and I don't think it makes sense to try to get that into TF 2.2.

My suggestion is to, optionally, submit an isolated fix for `prefetch_to_device` (which relies on the deprecated `colocate_with` API) and implement the proper fix in placer.

Commit:
Additional GPU Prefetch Fix

DEKHTIARJonathan commented on tensorflow/python/data/experimental/kernel_tests/prefetch_to_device_test.py:
```diff
@@ -146,11 +146,14 @@ def testPrefetchToDeviceGpu(self):
     host_dataset = dataset_ops.Dataset.range(10)
     device_dataset = host_dataset.apply(
-        prefetching_ops.prefetch_to_device("/gpu:0"))
+        prefetching_ops.prefetch_to_device('/gpu:0'))
```
Oops sorry my editor did this automatically ...

Commit:
Formatting fix tensorflow/python/data/experimental/kernel_tests/prefetch_to_device_test.py

DEKHTIARJonathan commented:
@jsimsa Exploring a little I find a bunch of couple identical bugs. I have updated the PR, however numerous changes needs to be done before/after merging this PR:

1. Relying on the TensorFlow placer implementation as you suggested instead of `ops.colocate_with()`.

2. Most TF Data Options are in fact incompatible with GPU placements. I removed the `.with_options()` from the `prefetch_to_device` implementation and directly bypassed the options in case we are running on GPU. This has two advantages:
    2.1 Users can still set the options ( even if they are not used ) and later benefits from any GPU implementation we provide with updating their code.
    2.2 This is more generic as most (all ?) of the optimizations are in fact failing if the `dataset._variant_tensor` is on the GPU.

@sanjoy Could we discuss a little the strategy here next Wednesday meeting ? @jsimsa feel free to join ;)

Commit:
Update prefetching_ops.py

Commit:
Update dataset_ops.py

Commit:
Update prefetching_ops.py

alextp commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -2194,14 +2194,17 @@ def _make_initializable_iterator(self, shared_name=None):  # pylint: disable=mis
       shared_name = ""
-    iterator_resource = gen_dataset_ops.iterator_v2(
+
+    with ops.colocate_with(self._variant_tensor):
```
In my experience colocate_with is very dangerous, and there's almost always a better way of achieving what we want than using colocate_with.

One problem is that `colocate_with(a): b` has different behavior in eager and graph modes: in eager mode it can only affect the placement of b, but in graph mode it also "retroactively" affects the placement of a as well. This is made doubly bad by the fact that colocation group sizes tend to grow out of control quickly (b is rarely a single op, it's often a bunch of ops, many with no particular reason to be there, and all these ops are now tied to the same colocation group, and any device placement constraints or kernel registration issues affecting one of them now affect all of them) and can have very weird behavior as well on multi-node clusters.

Given this I strongly prefer that we:
 1. ask users who are creating their own iterators from datasets to explicitly add device scopes to get the iterators on the right device; explicit here is much less buggy than implicit.
 2. change the distribution strategies which work on GPUs (like MirroredStrategy and the multi-worker equivalent) to add these device scopes, since at that point we know what the right user intent is
 3. figure out a way for horovod to signal user intent in a similar way as distribution strategy (or even implement a horovod strategy, for uniformity with the rest of TF, full keras compatibility, etc); it'll make it less magic.

In the long run I agree it'd be nice if we could manage to not lose information about the dataset when creating iterators; I don't know whether we want to propagate some metadata from python or if we want to propagate some metadata within the runtime and make it visible to the placer. Since this is unclear I'd prefer to defer this decision while unblocking people by supporting explicit placement.

Does that sound reasonable? I'm happy to jump into a videoconference call to clarify my position / listen to counter arguments.

DEKHTIARJonathan replied:
@alextp thanks you so much for you answer :) I wasn't aware of this issue with `colocate_with`.

I did something really naive and replaced `with ops.colocate_with()` to `with ops.device()`.
That actually perfectly works and even pass my unittest easy.

```bash
bazel test --verbose_failures //tensorflow/python/data/experimental/kernel_tests:prefetch_to_device_test
```

Could this be acceptable ?

it has the advantage to completely fix the colocation issue you raise and fully address point 1) and 2) without any additional change. And most likely will be good for Horovod (I gotta try).

Commit:
Change ops.colocate_with to ops.device

alextp commented on tensorflow/python/data/ops/iterator_ops.py:
```diff
@@ -367,7 +367,8 @@ def make_initializer(self, dataset, name=None):
                           (self.output_shapes, dataset_output_shapes))
-    with ops.colocate_with(self._iterator_resource):
+
+    with ops.colocate_with(dataset._variant_tensor):
```
Can you remove these colocations too?

DEKHTIARJonathan replied:
Done in 6fdaa56

alextp commented:
tf.device is much better, thanks. I am still unsure what happens if the
user places the iterator creation in a device scope; now we'll ignore their
action and always place it on the dataset device. I think the same number
of copies happens in either case but if the iterator gets put on the user's
device (as opposed to the device we choose) we might be able to a better
job of prefetching the tensors there.

I'm not sure; I need to think a little more about this.

On Tue, Apr 7, 2020 at 4:32 PM Jonathan DEKHTIAR <notifications@github.com>
wrote:

> *@DEKHTIARJonathan* commented on this pull request.
> ------------------------------
>
> In tensorflow/python/data/ops/dataset_ops.py
> <https://github.com/tensorflow/tensorflow/pull/37277#discussion_r405173524>
> :
>
> > @@ -2194,14 +2194,17 @@ def _make_initializable_iterator(self, shared_name=None):  # pylint: disable=mis
>      dataset = self._apply_options()
>      if shared_name is None:
>        shared_name = ""
> -    iterator_resource = gen_dataset_ops.iterator_v2(
> +
> +    with ops.colocate_with(self._variant_tensor):
>
> @alextp <https://github.com/alextp> thanks you so much for you answer :)
> I wasn't aware of this issue with colocate_with.
>
> I did something really naive and replaced with ops.colocate_with() to with
> ops.device().
> That actually perfectly works and even pass my unittest easy.
>
> bazel test --verbose_failures //tensorflow/python/data/experimental/kernel_tests:prefetch_to_device_test
>
> Could this be acceptable ?
>
> it has the advantage to completely fix the colocation issue you raise and
> fully address point 1) and 2) without any additional change. And most
> likely will be good for Horovod (I gotta try).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/37277#discussion_r405173524>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHROKTYE5Y32ZHOJXNA3RLOZ2TANCNFSM4LAYRPHA>
> .
>


-- 
 - Alex

Commit:
Change ops.colocate_with to ops.device in Iterator

Commit:
Change ops.colocate_with to ops.device in Iterator - 2

DEKHTIARJonathan commented:
Let's discuss tomorrow then. I'm sure any concern can be addressed. We seem on good track

jsimsa commented:
I agree that `tf.device` is better.  I am still not convinced that doing this in Python is better than doing this in placer but assuming the PR addresses my comments below I support the change.

The current version of the PR only fixes the iterator placement issue for TF 1 graph-mode only. 

Please extend this PR to also fix the iterator placement issue for TF 2 eager mode and tf.function graph-mode. This will require additional `tf.device` scopes and I will create a test that will help us validate that your change works in TF 2 eager mode and tf.function graph-mode. Furthermore, this PR should also use the device scope for `reduce`-based consumption of tf.data elements.

DEKHTIARJonathan commented:
@jsimsa do you want write access to my fork ? That way you can directly push your unittest in the PR.
When do you think we could have it ?

Great, sounds like a plan to me.

jsimsa commented:
I already created the test internally and will ping this PR when it is submitted.

jsimsa commented:
My change has been submitted and you will need to resolve merge conflicts.

On a second thought, I suggest that instead of updating existing unit tests, you write new unit tests that check that the `prefetch_with_device` and `iterator` op placement is fixed by your change. 

In other words, please add the following tests:

1) `prefetch_to_device_test.py`: eager mode and graph mode test that checks that the device of the prefetch dataset variant tensor matches the target device 

2) `iterator_test.py`: 
- eager mode test that checks that the device of the eagerly created iterator (i.e. `iter(dataset)` matches the device of the input dataset
- graph mode test that checks that the device of iterator created by `make_one_shot_iterator` and `make_initializable_iterator` matches the device of the input dataset

Commit:
Merge branch 'master' into patch-1

Commit:
Unittest fix

Commit:
Unittests modifications as requested

DEKHTIARJonathan commented:
@jsimsa @alextp I applied the requested changes, let me know if that looks good with you.

```bash
$ bazel test //tensorflow/python/data/experimental/kernel_tests:prefetch_to_device_test

Target //tensorflow/python/data/experimental/kernel_tests:prefetch_to_device_test up-to-date:
  bazel-bin/tensorflow/python/data/experimental/kernel_tests/prefetch_to_device_test
INFO: Elapsed time: 20.261s, Critical Path: 9.85s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
//tensorflow/python/data/experimental/kernel_tests:prefetch_to_device_test (cached) PASSED in 4.4s

INFO: Build completed successfully, 1 total action
```

```bash
$ bazel test //tensorflow/python/data/kernel_tests:iterator_test

INFO: Analyzed target //tensorflow/python/data/kernel_tests:iterator_test (328 packages loaded, 24880 targets configured).
INFO: Found 1 test target...
INFO: Deleting stale sandbox base /root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/sandbox
Target //tensorflow/python/data/kernel_tests:iterator_test up-to-date:
  bazel-bin/tensorflow/python/data/kernel_tests/iterator_test
INFO: Elapsed time: 27.329s, Critical Path: 17.28s
INFO: 1 process: 1 local.
INFO: Build completed successfully, 2 total actions
//tensorflow/python/data/kernel_tests:iterator_test                      PASSED in 7.1s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 2 total actions
```

jsimsa commented on tensorflow/python/data/kernel_tests/iterator_test.py:
```diff
@@ -1016,6 +1018,85 @@ def fn():
+
+    def assert_host_placement(_obj):
+      try:
+        self.assertIn("CPU:0", _obj)
```
You can avoid the try catch by doing:

`self.assertTrue("CPU:0" in host_dataset._variant_tensor.device or "" == host_dataset._variant_tensor.device)`

jsimsa reviewed (changes requested):
Thank you. The source code changes look good. Left a couple of comments on tests.

DEKHTIARJonathan commented on tensorflow/python/data/kernel_tests/iterator_test.py:
```diff
@@ -1016,6 +1018,85 @@ def fn():
+    self.assertIn("GPU:0", device_dataset._variant_tensor.device)
+    self.assertIn("GPU:0", device_tensor.device)
+
+    if not tf2.enabled() or context.executing_eagerly():
```
If I do check check device I obtain the error that the attribute `_device` does not exist in this scenario. Sounds like an issue to me but I didn't want to fix something disconnect from this PR.

jsimsa replied:
This is somewhat worrisome because if the test code cannot access `.device` that neither will non-test code and since the code you have modified will run in TF2 graph mode, I suspect that this is an issue.

@alextp does it make sense that the `.device` attribute is no present in TF2 graph mode?

alextp replied:
The Tensor.device attribute is present in tf2 graph mode. Or are we not getting the device of a tensor? Can you print the object which doesn't have the device attribute on a failure and paste the result here or in a gist?

Commit:
Changes Requested

Commit:
Bug Fix

DEKHTIARJonathan commented:
@jsimsa @alextp any update?

jsimsa commented:
@DEKHTIARJonathan Alex left a comment for you on Thursday that (AFAICT) you have not addressed:

> The Tensor.device attribute is present in tf2 graph mode. Or are we not getting the device of a tensor? Can you print the object which doesn't have the device attribute on a failure and paste the result here or in a gist?

DEKHTIARJonathan commented:
Hmm ... I did, anyway probably lost in the comment thread or dismissed by pushing an new commit.
By the way, here the issue is the iterator (not a tensor) missing `_device` attribute (not `device`):

----------------

So by just commenting `if not tf2.enabled() or context.executing_eagerly():`

I obtain the following error:

```python
======================================================================
ERROR: testIteratorOnDeviceGraphModeInitializableIterator_test_mode_graph_tfapiversion_2 (__main__.IteratorTest)
testIteratorOnDeviceGraphModeInitializableIterator_test_mode_graph_tfapiversion_2 (__main__.IteratorTest)
testIteratorOnDeviceGraphModeInitializableIterator_test_mode_graph_tfapiversion_2(mode='graph', tf_api_version=2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/absl_py/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 314, in decorated
    execute_test_method()
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 297, in execute_test_method
    test_method(**kwargs_to_pass)
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/iterator_test.py", line 1098, in testIteratorOnDeviceGraphModeInitializableIterator
    device_dataset, device_iterator, device_tensor
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/iterator_test.py", line 1037, in assert_dataset_placement
    "cpu:0" in host_iterator._device.lower() or host_iterator._device == ""
AttributeError: 'Iterator' object has no attribute '_device'

======================================================================
ERROR: testIteratorOnDeviceGraphModeOneShotIterator_test_mode_graph_tfapiversion_2 (__main__.IteratorTest)
testIteratorOnDeviceGraphModeOneShotIterator_test_mode_graph_tfapiversion_2 (__main__.IteratorTest)
testIteratorOnDeviceGraphModeOneShotIterator_test_mode_graph_tfapiversion_2(mode='graph', tf_api_version=2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/absl_py/absl/testing/parameterized.py", line 263, in bound_param_test
    test_method(self, **testcase_params)
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 314, in decorated
    execute_test_method()
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 297, in execute_test_method
    test_method(**kwargs_to_pass)
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/iterator_test.py", line 1078, in testIteratorOnDeviceGraphModeOneShotIterator
    device_dataset, device_iterator, device_tensor
  File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/iterator_test.py", line 1037, in assert_dataset_placement
    "cpu:0" in host_iterator._device.lower() or host_iterator._device == ""
AttributeError: 'Iterator' object has no attribute '_device'

----------------------------------------------------------------------
```

```python
>>> dir(host_iterator)

['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_add_variable_with_custom_getter', '_checkpoint_dependencies', '_deferred_dependencies', '_element_spec', '_flat_tensor_shapes', '_flat_tensor_types', '_gather_saveables_for_checkpoint', '_get_next_call_count', '_handle_deferred_dependencies', '_initializer', '_iterator_resource', '_list_extra_dependencies_for_serialization', '_list_functions_for_serialization', '_lookup_dependency', '_maybe_initialize_trackable', '_name_based_attribute_restore', '_name_based_restores', '_no_dependency', '_object_identifier', '_preload_simple_restoration', '_restore_from_checkpoint_position', '_setattr_tracking', '_single_restoration_from_checkpoint_position', '_string_handle', '_tf_api_names', '_tf_api_names_v1', '_track_trackable', '_tracking_metadata', '_unconditional_checkpoint_dependencies', '_unconditional_dependency_names', '_update_uid', 'element_spec', 'from_string_handle', 'from_structure', 'get_next', 'initializer', 'make_initializer', 'output_classes', 'output_shapes', 'output_types', 'string_handle']
```

As you can see `_device` does not exist

```python
>>> import pprint
>>> pprint.pprint(host_iterator.__dict__, indent=4)

{  
    '_element_spec': TensorSpec(shape=(), dtype=tf.int64, name=None),
    '_flat_tensor_shapes': [TensorShape([])],
    '_flat_tensor_types': [tf.int64],
    '_get_next_call_count': 1, 
    '_initializer': <tf.Operation 'MakeIterator' type=MakeIterator>,
    '_iterator_resource': <tf.Tensor 'IteratorV2:0' shape=() dtype=resource>,
    '_string_handle': <tf.Tensor 'IteratorToStringHandle:0' shape=() dtype=string>
}
```

alextp commented:
It needs to be iterator._iterator_resource.device, it seems

On Mon, Apr 13, 2020 at 9:45 AM Jonathan DEKHTIAR <notifications@github.com>
wrote:

> Hmm ... I did, anyway probably lost in the comment thread or dismissed by
> pushing an new commit.
> By the way, here the issue is the iterator (not a tensor) missing _device
> attribute (not device):
> ------------------------------
>
> So by just commenting if not tf2.enabled() or context.executing_eagerly():
>
> I obtain the following error:
>
> ======================================================================ERROR: testIteratorOnDeviceGraphModeInitializableIterator_test_mode_graph_tfapiversion_2 (__main__.IteratorTest)
> testIteratorOnDeviceGraphModeInitializableIterator_test_mode_graph_tfapiversion_2 (__main__.IteratorTest)
> testIteratorOnDeviceGraphModeInitializableIterator_test_mode_graph_tfapiversion_2(mode='graph', tf_api_version=2)----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/absl_py/absl/testing/parameterized.py", line 263, in bound_param_test
>     test_method(self, **testcase_params)
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 314, in decorated
>     execute_test_method()
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 297, in execute_test_method
>     test_method(**kwargs_to_pass)
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/iterator_test.py", line 1098, in testIteratorOnDeviceGraphModeInitializableIterator
>     device_dataset, device_iterator, device_tensor
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/iterator_test.py", line 1037, in assert_dataset_placement
>     "cpu:0" in host_iterator._device.lower() or host_iterator._device == ""AttributeError: 'Iterator' object has no attribute '_device'
> ======================================================================ERROR: testIteratorOnDeviceGraphModeOneShotIterator_test_mode_graph_tfapiversion_2 (__main__.IteratorTest)
> testIteratorOnDeviceGraphModeOneShotIterator_test_mode_graph_tfapiversion_2 (__main__.IteratorTest)
> testIteratorOnDeviceGraphModeOneShotIterator_test_mode_graph_tfapiversion_2(mode='graph', tf_api_version=2)----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/absl_py/absl/testing/parameterized.py", line 263, in bound_param_test
>     test_method(self, **testcase_params)
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 314, in decorated
>     execute_test_method()
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/framework/test_combinations.py", line 297, in execute_test_method
>     test_method(**kwargs_to_pass)
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/iterator_test.py", line 1078, in testIteratorOnDeviceGraphModeOneShotIterator
>     device_dataset, device_iterator, device_tensor
>   File "/root/.cache/bazel/_bazel_root/a8fc6d0749b4f3c43761726a36e8ec4c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/data/kernel_tests/iterator_test.runfiles/org_tensorflow/tensorflow/python/data/kernel_tests/iterator_test.py", line 1037, in assert_dataset_placement
>     "cpu:0" in host_iterator._device.lower() or host_iterator._device == ""AttributeError: 'Iterator' object has no attribute '_device'
> ----------------------------------------------------------------------
>
> >>> dir(host_iterator)
>
> ['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_add_variable_with_custom_getter', '_checkpoint_dependencies', '_deferred_dependencies', '_element_spec', '_flat_tensor_shapes', '_flat_tensor_types', '_gather_saveables_for_checkpoint', '_get_next_call_count', '_handle_deferred_dependencies', '_initializer', '_iterator_resource', '_list_extra_dependencies_for_serialization', '_list_functions_for_serialization', '_lookup_dependency', '_maybe_initialize_trackable', '_name_based_attribute_restore', '_name_based_restores', '_no_dependency', '_object_identifier', '_preload_simple_restoration', '_restore_from_checkpoint_position', '_setattr_tracking', '_single_restoration_from_checkpoint_position', '_string_handle', '_tf_api_names', '_tf_api_names_v1', '_track_trackable', '_tracking_metadata', '_unconditional_checkpoint_dependencies', '_unconditional_dependency_names', '_update_uid', 'element_spec', 'from_string_handle', 'from_structure', 'get_next', 'initializer', 'make_initializer', 'output_classes', 'output_shapes', 'output_types', 'string_handle']
>
> As you can see _device does not exist
>
> >>> import pprint>>> pprint.pprint(host_iterator.__dict__, indent=4)
>
> {
>     '_element_spec': TensorSpec(shape=(), dtype=tf.int64, name=None),
>     '_flat_tensor_shapes': [TensorShape([])],
>     '_flat_tensor_types': [tf.int64],
>     '_get_next_call_count': 1,
>     '_initializer': <tf.Operation 'MakeIterator' type=MakeIterator>,
>     '_iterator_resource': <tf.Tensor 'IteratorV2:0' shape=() dtype=resource>,
>     '_string_handle': <tf.Tensor 'IteratorToStringHandle:0' shape=() dtype=string>
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/37277#issuecomment-612982080>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRJSPIMZZDRWNQ5OPYTRMM6TLANCNFSM4LAYRPHA>
> .
>


-- 
 - Alex

DEKHTIARJonathan commented:
@alextp Funny enough, what I showed is working perfectly in TF1 or TF2 Eager.
Shall I just update with you propose ?

DEKHTIARJonathan commented:
@alextp are you sure replacing with `iterator._iterator_resource.device` is the proper way ?
As stated above: `host_iterator._device` is working perfectly in TF1 or TF2 Eager.

alextp commented:
Oh if it works in tf1 and tf2 then it's fine.

On Wed, Apr 22, 2020 at 1:53 PM Jonathan DEKHTIAR <notifications@github.com>
wrote:

> @alextp <https://github.com/alextp> are you sure replacing with
> iterator._iterator_resource.device is the proper way ?
> As stated above: host_iterator._device is working perfectly in TF1 or TF2
> Eager.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/37277#issuecomment-618034617>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRI52P6LDXL7U3WWHIDRN5KNFANCNFSM4LAYRPHA>
> .
>


-- 
 - Alex

DEKHTIARJonathan commented:
It just doesn't work in TF2 graph mode. The attribute does not exist. Hence I only check this for TF1 and TF2 eager. If that's good with you then we can merge

alextp commented:
It also needs to work in tf2 graph mode.

On Wed, Apr 22, 2020 at 2:02 PM Jonathan DEKHTIAR <notifications@github.com>
wrote:

> It just doesn't work in TF2 graph mode. The attribute does not exist.
> Hence I only check this for TF1 and TF2 eager. If that's good with you then
> we can merge
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/37277#issuecomment-618038688>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRKZEX733EEU4AMCD5DRN5LOFANCNFSM4LAYRPHA>
> .
>


-- 
 - Alex

DEKHTIARJonathan commented:
I can't fix it in this PR. It looks to be an issue in the Iterator in TF2 graph mode. A complete different issue than this PR tries to solve.

alextp commented:
Can't you do the fix I suggested if in tf2 graph mode? Something like if
not executing_eagerly() and executing_eagerly_outside_functions()?

On Wed, Apr 22, 2020 at 2:05 PM Jonathan DEKHTIAR <notifications@github.com>
wrote:

> I can't fix it in this PR. It looks to be an issue in the Iterator in TF2
> graph mode. A complete different issue than this PR tries to solve.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/37277#issuecomment-618040158>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRPWA4Y7XGHCHXGDOOTRN5L2JANCNFSM4LAYRPHA>
> .
>


-- 
 - Alex

DEKHTIARJonathan commented:
That's exactly what I did: https://github.com/tensorflow/tensorflow/blob/65b1a97ab98f09b197ee7b06a1ac6290349be816/tensorflow/python/data/kernel_tests/iterator_test.py#L1036-L1038

@jsimsa was complaining about this:
> Why are are we not checking the iterator device placement in TF2 graph mode?
> This is somewhat worrisome because if the test code cannot access .device that neither will non-test code and since the code you have modified will run in TF2 graph mode, I suspect that this is an issue.
> 
> @alextp does it make sense that the .device attribute is no present in TF2 graph mode?

jsimsa commented:
The reason you should use `_iterator_resource` instead of `_device` is because there are two different implementations of tf.data.Iterator -- [one](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/ops/iterator_ops.py#L546) for eager mode + tf.function and [one](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/ops/iterator_ops.py#L77) for graph mode. Only the former has the `_device` property, while both of them have the `_iterator_resource` property.

Commit:
Unittest Changes requested

DEKHTIARJonathan commented:
Perfect. Just wanted to make sure this was expected and we understood correctly the issue ;).
Just pushed the fix. Let me know if there's any other blocking issue to merge this PR.

```bash
Target //tensorflow/python/data/experimental/kernel_tests:prefetch_to_device_test up-to-date:
  bazel-bin/tensorflow/python/data/experimental/kernel_tests/prefetch_to_device_test
INFO: Elapsed time: 19.913s, Critical Path: 6.06s
INFO: 1 process: 1 local.
INFO: Build completed successfully, 5 total actions
//tensorflow/python/data/experimental/kernel_tests:prefetch_to_device_test PASSED in 3.8s
```

```bash
Target //tensorflow/python/data/kernel_tests:iterator_test up-to-date:
  bazel-bin/tensorflow/python/data/kernel_tests/iterator_test
INFO: Elapsed time: 14.175s, Critical Path: 1.97s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
//tensorflow/python/data/kernel_tests:iterator_test             (cached) PASSED in 5.0s
```

jsimsa commented on tensorflow/python/data/ops/iterator_ops.py:
```diff
@@ -419,13 +420,14 @@ def get_next(self, name=None):
+    with ops.device(self._iterator_resource.device):
+      # pylint: disable=protected-access
+      flat_ret = gen_dataset_ops.iterator_get_next(
         self._iterator_resource,
```
you will need to fix the indent here

jsimsa replied:
these used to be indented by 4 spaces from the previous line before your change and now seem to be indented by 2 spaces

jsimsa replied:
You can't find all the lint warnings in https://source.cloud.google.com/results/invocations/eae2a68e-6a74-41aa-8e87-dfd6855e6249/log

DEKHTIARJonathan commented:
@jsimsa https://github.com/tensorflow/tensorflow/pull/37277#discussion_r413351911

I can't see the issue with the indent here. Looks good to me ...

Commit:
Linting Issues Fix

DEKHTIARJonathan commented:
@jsimsa fixed linting issue: 0f26cfa90c2477392da1b1409f690c70cdf050f7
Can you re-approve the PR ? Thanks

jsimsa commented:
There is still a couple of lint issues:

```
FAIL: Found 3 non-whitelisted pylint errors:
tensorflow/python/data/ops/dataset_ops.py:2277: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).

tensorflow/python/data/ops/dataset_ops.py:2285: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).

tensorflow/python/data/ops/dataset_ops.py:2286: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
```

AFAICT, the windows failures are unrelated.

DEKHTIARJonathan commented:
@jsimsa I did not modify these lines
https://github.com/tensorflow/tensorflow/blob/0f26cfa90c2477392da1b1409f690c70cdf050f7/tensorflow/python/data/ops/dataset_ops.py#L2277-L2287

And honestly I'm not sure how to fix them, if I apply the suggested fix, i'll get indent errors.

jsimsa commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -2252,14 +2258,17 @@ def _make_initializable_iterator(self, shared_name=None):  # pylint: disable=mis
+
+    with ops.device(self._variant_tensor.device):
+      iterator_resource = gen_dataset_ops.iterator_v2(
         container="", shared_name=shared_name, **self._flat_structure)
```
wrong indent here

jsimsa commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -2252,14 +2258,17 @@ def _make_initializable_iterator(self, shared_name=None):  # pylint: disable=mis
+
+      # pylint: disable=protected-access
+      return iterator_ops.Iterator(
         iterator_resource, initializer, get_legacy_output_types(dataset),
```
wrong indent here

jsimsa commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -2252,14 +2258,17 @@ def _make_initializable_iterator(self, shared_name=None):  # pylint: disable=mis
+      # pylint: disable=protected-access
+      return iterator_ops.Iterator(
         iterator_resource, initializer, get_legacy_output_types(dataset),
         get_legacy_output_shapes(dataset), get_legacy_output_classes(dataset))
```
wrong indent here

jsimsa commented:
> @jsimsa I did not modify these lines
> https://github.com/tensorflow/tensorflow/blob/0f26cfa90c2477392da1b1409f690c70cdf050f7/tensorflow/python/data/ops/dataset_ops.py#L2275-L2287
> 
> And honestly I'm not sure how to fix them, if I apply the suggested fix, i'll get indent errors.

The line numbers are wrong (not sure why) but the indent issues exists. You could have just carefully review the PR diff. To make it easier for you, I commented on the lines that have the wrong indent in your PR.

Commit:
Linting Issues Fix

DEKHTIARJonathan commented:
@jsimsa linting issues fixed in 356a3dd 
Please approve again ;)

Any way I can run the linter by myself ?

jsimsa commented:
FYI, this got rolled back automatically because of internal test failures. I will take a look, time permitting. Please ping me if you do not see a response in a week.

Flamefire commented:
@jsimsa *ping*

DEKHTIARJonathan commented:
@jsimsa any chance we could get an update on when do you plan to re-merge this PR ?  It's a quite important PR that impact performance. Let me know if I can help.

aaudiber commented:
This should be fixed now by https://github.com/tensorflow/tensorflow/commit/8be4d61574f29568c8699708d88945b441bfd317

DEKHTIARJonathan commented:
@aaudiber : the commit you linked is far from addressing all concerns. If you look at this PR diff, you will see that it modifies much more things.

It namely completely ignores placing the Iterator on the correct device. Consequently, doing `x = iterator.get_next()` triggers a copy back on CPU. Unfortunately, this makes 8be4d61 totally useless since we bump in the same issue a bit further down the input pipeline

CC: @nluehr @jsimsa

aaudiber commented:
@DEKHTIARJonathan The linked commit fixes a bug where the prefetch dataset would be placed on CPU. As demonstrated by the included unit test, the prefetch dataset is now placed on GPU.

There is a separate issue that iterators are not automatically colocated with their inputs. We originally tried to address both issues with this PR, but automatic colocation broke internal tests, so we had to roll this PR back. It will take some time to figure out the root issue. It should be possible to work around the issue by explicitly putting the `get_next` inside a device scope. Could you open a separate issue for iterator colocation? Please include details about why it is important, to help us prioritize the issue.
